### PR TITLE
Random shuffle for skewed joins optimization

### PR DESCRIPTION
**Pull request by mlyublena**
Randomize the skewed side of a join by transforming:
```
      join (REPLICATED)
        - scan S
        - scan T
 ```
into
 ```
      join (REPLICATED)
          Exchange(partition by (random))
            Project(random=random(num_shards))
              - scan S
        - scan T
```

Note: this is the base of the optimization, will implement skew detection through HBO in separate commits.

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ShardedJoin.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+ * into
+ *
+ *     join (REPLICATED)
+ *       Filter (rownum>=1)
+ *         Rownumber(partition by (random))
+ *           Project(random=random(num_shards))
+ *             - scan S
+ *       - scan T
+ *
+ */
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ShardedJoin.java:71]: Maybe I lost some context, but the rewrite seems not the sharded join we were talking about? cc @kaikalur

Comment (mlyublena) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ShardedJoin.java:71]: discussed offline with @kaikalur , indeed this is randomize shuffle on probe side, will rename/simplify implementation as per discussion

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomShuffleProbeSide.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        private static boolean isApplicable(JoinNode joinNode, FeaturesConfig.RandomizeProbeSideStrategy strategy)
+        {
+            return isBroadcastJoin(joinNode) &&
+                    (strategy == FeaturesConfig.RandomizeProbeSideStrategy.ALWAYS ||
+                            strategy == FeaturesConfig.RandomizeProbeSideStrategy.COST_BASED && isLeftSideSkewed(joinNode));
+        }
+
+        private static boolean isLeftSideSkewed(JoinNode joinNode)
```
Comment (agrawaldevesh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomShuffleProbeSide.java:128]: Curious and just trying to learn so a n00b q: will Hbo also be giving us what keys are skewed ?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestRandomShuffleProbeSide.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Test
+    public void testJoin()
+    {
+        assertPlan("SELECT * FROM orders JOIN lineitem ON orders.orderkey = lineitem.orderkey",
+                getSessionAlwaysEnabled(),
+                anyTree(
+                        join(
+                                INNER,
+                                ImmutableList.of(equiJoinClause("leftOrderKey", "rightOrderKey")),
```
Comment (agrawaldevesh) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestRandomShuffleProbeSide.java:55]: Again trying to learn: where is the check for random here ?

**Review by agrawaldevesh — COMMENTED** [review]
Really interesting optimization :-) would love to read a design doc or sketch showing the whole flow with hbo in the mix.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestRandomShuffleProbeSide.java (hunk 3 above)
Comment (mlyublena) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestRandomShuffleProbeSide.java:55]: @agrawaldevesh , this is an old code review (forgot to close it), please check the new version here: https://github.com/prestodb/presto/pull/21037

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomShuffleProbeSide.java (hunk 2 above)
Comment (mlyublena) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomShuffleProbeSide.java:128]: we don't plan to store skewed key values at the moment due to privacy implications regarding user data; rather we will store flags whether the data is skewed or not

**Comment by mlyublena** [comment]
New implementation here: https://github.com/prestodb/presto/pull/21037

**Pull request closed without merging**